### ISSUE TITLE
yasnippet breaks iedit-mode

### ISSUE THREAD

**Issue by netjune** [joaotavora/yasnippet#1041]
When iedit-mode is on, if you expand a snippet by yas-expand-snippet, iedit-mode will be breaked. Because it depends on inhibit-modification-hooks to be nil to run its hook. But yasnippet bind it to t when it creates snippet. So iedit-mode can't work.

**Comment by npostavs** [comment]
It's more complicated than just `inhibit-modification-hooks. `yas--snippet-create` already does `(run-hook-with-args 'before-change-functions begin begin)` and `(run-hook-with-args 'after-change-functions (point-min) (point-max) 0)`, so that part should be covered.

But how should these two packages interact together? I guess if you expand a snippet in one mode then you get a copy in each iedit region, but yasnippet can't handle multiple snippets at a time, so it's not very easy to make this work.

**Comment by netjune** [comment]
In fact it is an empty snippet. It is only a simple string. There are no snippet fields in it. company-lsp always calls yas-expand to try to expand the completion item as a snippet even it is not.
After yas-expand returns, there is no active snippet in buffer.

**Comment by npostavs** [comment]
Oh, the problem is that `inhibit-modification-hooks` blocks overlay
modification hooks too, which we don't call outselves.  Does the patch
in #1046 help?

**Comment by netjune** [comment]
Yes. It works now. Thanks.

**Comment by netjune** [comment]
A related question: Is it safe for yasnippet to abort when there are multiple active snippets? It may happens by accidently.

**Comment by npostavs** [comment]
Thanks for testing.

> Is it safe for yasnippet to abort when there are multiple active snippets? It may happens by accident.

I'm not sure what you mean.

**Comment by netjune** [comment]
When iedit-mode is active, if a snippet with more than one fields is expanded by accident, there will be multiple active snippets and buffer text is messed. Is there a safe way to avoid this?

**Comment by npostavs** [comment]
> When iedit-mode is active, if a snippet with more than one fields is expanded by accident, there will be multiple active snippets and buffer text is messed.

Actually, I think iedit would copy only the text, not the yasnippet field structures, so snippets won't be duplicated.

**Comment by netjune** [comment]
> > When iedit-mode is active, if a snippet with more than one fields is expanded by accident, there will be multiple active snippets and buffer text is messed.
> 
> Actually, I think iedit would copy only the text, not the yasnippet field structures, so snippets won't be duplicated.

It seems that it works sometimes but not always.

**Comment by netjune** [comment]
Another problem: The undo list seems be broken.

**Comment by npostavs** [comment]
Can you show an example?

**Comment by netjune** [comment]
I uploaded an screencast here: https://github.com/netjune/temp_files/blob/master/yas-bug1.mp4

**Comment by npostavs** [comment]
I find that hard to follow. What is the snippet? It looks more like a company-mode expansion?

**Comment by netjune** [comment]
Yes. It is a company-mode expansion. The snippet has no fields. So "vbox_map" ==> "vbox_map"

**Comment by netjune** [comment]
https://github.com/joaotavora/yasnippet/blob/5b1217ab085fab4abeb1118dccb260691b446703/yasnippet.el#L4171

Is undo is disabled in this block? Then iedit may can't push its undo info to undo list.

**Comment by npostavs** [comment]
Good point, does the patch in #1051 help?

**Comment by netjune** [comment]
> Good point, does the patch in #1051 help?

I tried. It doesn't help.

It seems that with the above patch iedit-mode still do its work in the block `(let ((buffer-undo-list t)) ` while it can't record its undo info.

Maybe we can create the snippet in a temp buffer and then copy the finnal text to the original buffer. Then there will be only one text change and we don't have to disable undo and it will not break iedit-mode. But we can't indent the snippet in a temp buffer. 

I think this bug is too hard to fix.

**Comment by npostavs** [comment]
joaotavora recently fixed a bug I'd made in #1046: https://github.com/joaotavora/yasnippet/commit/d3d6d70b1cd4818d271752468e0fdb0788db750d

It might fix your problem with iedit too.

**Comment by netjune** [comment]
Today, I tried https://raw.githubusercontent.com/joaotavora/yasnippet/321c3bbdcbebba33e3878aeed2b08e85e311b859/yasnippet.el. It is worse than before. It inserts unexpected text after expanding. And it also breaks undo list.

**Comment by sebastiencs** [comment]
Hello,

I also have issues on the master branch.
After inserting a snippet:
- all the buffer get indented
- my point move to the end of the buffer
- `undo` is broken

```diff
@@ -4209,8 +4209,8 @@ Returns the newly created snippet."
         (setf (yas--snippet-control-overlay snippet)
               (yas--make-control-overlay snippet (point-min) (point-max)))
 
-        ;; Move to end
-        (goto-char (point-max))
+        ;; ;; Move to end
+        ;; (goto-char (point-max))
 
         (push snippet yas--active-snippets)
         snippet))))
```
The patch above fix the cursor position.
I have to set `(setq yas-indent-line nil)` to fix the indentation.
`undo` is still broken (the buffer becomes empty)
